File: sample/src/main/java/me/xiaopan/sketchsample/adapter/itemfactory/StaggeredImageItemFactory.java
package me.xiaopan.sketchsample.adapter.itemfactory;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import butterknife.BindView;
import me.xiaopan.assemblyadapter.AssemblyRecyclerItemFactory;
import me.xiaopan.sketch.util.SketchUtils;
import me.xiaopan.sketchsample.ImageOptions;
import me.xiaopan.sketchsample.R;
import me.xiaopan.sketchsample.adapter.BindAssemblyRecyclerItem;
import me.xiaopan.sketchsample.bean.BaiduImage;
import me.xiaopan.sketchsample.widget.SampleImageView;

public class StaggeredImageItemFactory extends AssemblyRecyclerItemFactory<StaggeredImageItemFactory.StaggeredImageItem> {
    private OnItemClickListener onItemClickListener;
    private int itemWidth;

    public StaggeredImageItemFactory(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    @Override
    public boolean isTarget(Object o) {
        return o instanceof BaiduImage;
    }

    @Override
    public StaggeredImageItem createAssemblyItem(ViewGroup viewGroup) {
        return new StaggeredImageItem(R.layout.list_item_image_staggered, viewGroup);
    }

    public interface OnItemClickListener {
        void onItemClick(int position, BaiduImage image, String loadingImageOptionsInfo);
    }

    public class StaggeredImageItem extends BindAssemblyRecyclerItem<BaiduImage> {
        @BindView(R.id.image_staggeredImageItem)
        SampleImageView imageView;

        public StaggeredImageItem(int itemLayoutId, ViewGroup parent) {
            super(itemLayoutId, parent);
        }

        @Override
        protected void onConfigViews(Context context) {
            imageView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (onItemClickListener != null) {
                        onItemClickListener.onItemClick(getAdapterPosition(), getData(), imageView.getOptionsKey());
                    }
                }
            });
            imageView.setOptions(ImageOptions.RECT);

            imageView.setPage(SampleImageView.Page.SEARCH_LIST);
        }

        @Override
        protected void onSetData(int i, BaiduImage image) {
            if (itemWidth == 0) {
                int screenWidth = imageView.getContext().getResources().getDisplayMetrics().widthPixels;
                itemWidth = (screenWidth - (SketchUtils.dp2px(imageView.getContext(), 4) * 3)) / 2;
            }

            ViewGroup.LayoutParams headParams = imageView.getLayoutParams();
            headParams.width = itemWidth;
            headParams.height = (int) (itemWidth / (image.getWidth() / (float) image.getHeight()));
            imageView.setLayoutParams(headParams);

            imageView.displayImage(image.getSourceUrl());
        }
    }
}

File: sample/src/main/java/me/panpf/sketch/sample/adapter/itemfactory/StaggeredImageItemFactory.kt
package me.panpf.sketch.sample.adapter.itemfactory

import android.content.Context
import android.view.View
import android.view.ViewGroup
import com.google.android.flexbox.FlexboxLayoutManager
import me.panpf.sketch.sample.ImageOptions
import me.panpf.sketch.sample.R
import me.panpf.sketch.sample.bean.BaiduImage
import me.panpf.sketch.sample.bindView
import me.panpf.sketch.sample.kotlinextends.isPortraitOrientation
import me.panpf.sketch.sample.widget.SampleImageView
import me.panpf.adapter.AssemblyItem
import me.panpf.adapter.AssemblyItemFactory

class StaggeredImageItemFactory(private val onItemClickListener: OnItemClickListener?) : AssemblyItemFactory<StaggeredImageItemFactory.StaggeredImageItem>() {

    override fun isTarget(o: Any): Boolean {
        return o is BaiduImage
    }

    override fun createAssemblyItem(viewGroup: ViewGroup): StaggeredImageItem {
        return StaggeredImageItem(R.layout.list_item_image_staggered, viewGroup)
    }

    interface OnItemClickListener {
        fun onItemClick(position: Int, image: BaiduImage, loadingImageOptionsInfo: String)
    }

    inner class StaggeredImageItem(itemLayoutId: Int, parent: ViewGroup) : AssemblyItem<BaiduImage>(itemLayoutId, parent) {
        val imageView: SampleImageView by bindView(R.id.image_staggeredImageItem)

        override fun onConfigViews(context: Context) {
            imageView.onClickListener = View.OnClickListener {
                onItemClickListener?.onItemClick(adapterPosition, data, imageView.optionsKey)
            }
            imageView.setOptions(ImageOptions.RECT)

            imageView.page = SampleImageView.Page.SEARCH_LIST

            imageView.layoutParams?.let {
                if (it is FlexboxLayoutManager.LayoutParams) {
                    it.flexGrow = 1.0f
                }
            }
        }

        override fun onSetData(i: Int, image: BaiduImage) {
            imageView.layoutParams?.let {
                it.height = imageView.context.resources.displayMetrics.heightPixels / (if (imageView.context.isPortraitOrientation()) 5 else 2)
                it.width = (it.height / (image.height / image.width.toFloat())).toInt()
                imageView.layoutParams = it
            }

            imageView.displayImage(image.url ?: "")
        }
    }
}
